src/bluesky/simulators.py

```diff
@@ -102,5 +117,285 @@ async def check_limits_async(plan):
+        In order to exercise a plan containing the following:
+
+        >>> def trigger_and_return_pin_tip():
+        ...     yield from bps.trigger(pin_tip)
+        ...     tip_pos = yield from bps.rd(pin_tip.triggered_tip)
+        ...     return tip_pos
+
+        >>> sim = RunEngineSimulator()
+        >>> sim.add_read_handler("pin_tip_detection-triggered_tip", (8, 5))
+        >>> msgs = sim.simulate_plan(trigger_and_return_pin_tip())
+        >>> assert sim.return_value == (8, 5)
+        """
+
+        def handle_read(_):
+            return value if isinstance(value, dict) else {dict_key: Reading(value=value, timestamp=time())}
+
+        self.add_handler("read", handle_read, obj_name)
+
+    def add_wait_handler(self, handler: Callable[[Msg], None], group: str = GROUP_ANY) -> None:
+        """Add a wait handler for a particular message.
+
+        Parameters
+        ----------
+        handler
+            a lambda that accepts a Msg, use this to execute any code that simulates something that's
+            supposed to complete when a group finishes
+        group
+            name of the group to wait for, default is GROUP_ANY which matches them all
+
+        Examples
+        --------
+        In order to exercise the following plan:
+        >>> def close_shutter():
+        ...     yield from bps.abs_set(detector_motion.shutter, 1, group="my_group")
+        ...     yield from bps.wait("my_group")
+        ...     done = yield from bps.rd("detector_motion_shutter")
+        ...     assert done == 1
+
+        >>> sim = RunEngineSimulator()
+        ... def simulate_detector_motion():
+        ...     sim.add_read_handler("detector_motion_shutter", 1)
+        ...
+        ... sim.add_wait_handler(simulate_detector_motion, "my_group")
+        ... sim.simulate_plan(close_shutter())
+        """
+        self.add_handler(
+            "wait",
+            handler,
+            lambda msg: (group == RunEngineSimulator.GROUP_ANY or msg.kwargs["group"] == group),
+        )
+
+    def add_callback_handler_for(
+        self,
+        command: str,
+        *,
+        document_name: Optional[str] = None,
+        document: Optional[dict] = None,
+        docs: Optional[Sequence[Sequence[Tuple[str, dict]]]] = None,
+        msg_filter: Optional[Callable[[Msg], bool]] = None,
+    ):
+        """Add a handler to fire one or more callbacks in sequence when a matching command is encountered.
+
+        To fire a single callback just once, specify `document_name` and `document` parameters.
+        To fire a sequence of callbacks, across one or more message invocations, use the `docs` parameter.
+
+        Parameters
+        ----------
+        command
+            Name of the command to match.
+        document_name
+            Name of the document to match when used in single shot mode.
+        document
+            The document to fire in single shot mode.
+        docs: Sequence of Sequence of document_name, document tuples
```

Comment:
Again, having two different call patterns squashed into one signature is a less than great pattern.

This might be better as two methods (one of which does indeed call the second with two layers of nesting).  I was going to suggest `add_callback_handler_for(..., *namedoc, msg_filter=None)` but the second level of nesting makes that not work.